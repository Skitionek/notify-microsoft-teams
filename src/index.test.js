const { run } = require('./index');

// Mock the core module
jest.mock('@actions/core');
const core = require('@actions/core');

// Mock the process.env object
process.env = {};

const MSTeams = require('./MSTeams');


describe('run function', () => {
    const defaultParams = {
        webhook_url: 'dummy_webhook',
        job: JSON.stringify({ id: 'job_1', status: 'success' }),
        steps: JSON.stringify([{ name: 'step_1', status: 'success' }]),
        needs: JSON.stringify({ need_1: { result: 'success' } }),
        title: 'Test Title',
        msteams_emails: 'test@example.com',
        raw: '',
        dry_run: 'false',
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should send a notification with the correct payload', async () => {
        core.getInput.mockImplementation((name) => defaultParams[name]);
        const mockGeneratePayload = jest.spyOn(MSTeams.prototype, 'generatePayload').mockImplementation(jest.fn());
        
        mockGeneratePayload.mockResolvedValue({ message: 'payload' });

        await run();

        expect(mockGeneratePayload).toHaveBeenCalledTimes(1);

        expect(mockGeneratePayload).toHaveBeenCalledWith(expect.any(Object));
    });
 
    it('should handle invalid JSON in inputs gracefully', async () => {
        core.getInput.mockImplementation((name) => {
            if (name === 'job') return '{invalid_json';
            return '{}';
        });
        const mockSetFailed = jest.spyOn(core, 'setFailed');

        await run();

        expect(mockSetFailed).toHaveBeenCalled(            expect.string        );
    });

    it('should handle empty msteams_emails gracefully', async () => {
        const params = {
            ...defaultParams,
            msteams_emails: ''
        };
        core.getInput.mockImplementation((name) => params[name]);
        const mockNotify = jest.spyOn(MSTeams.prototype, 'notify').mockImplementation(jest.fn());

        await run();

        expect(mockNotify).toHaveBeenCalledWith('dummy_webhook', expect.any(Object));
    });

    it('should handle missing title gracefully', async () => {
        const params = {
            ...defaultParams,
            title: ''
        };
        core.getInput.mockImplementation((name) => params[name]);
        const mockNotify = jest.spyOn(MSTeams.prototype, 'notify').mockImplementation(jest.fn());

        await run();

        expect(mockNotify).toHaveBeenCalledWith('dummy_webhook', expect.any(Object));
    });

    it('should handle errors in generatePayload', async () => {
        const mockGeneratePayload = jest.spyOn(MSTeams.prototype, 'generatePayload').mockImplementation(jest.fn());
        mockGeneratePayload.mockRejectedValue(new Error('Payload generation failed'));

        const mockSetFailed = jest.spyOn(core, 'setFailed');

        await run();

        expect(mockSetFailed).toHaveBeenCalledWith('Payload generation failed');
    });

    it('should handle errors in notify', async () => {
        const mockNotify = jest.spyOn(MSTeams.prototype, 'notify').mockImplementation(jest.fn());
        mockNotify.mockRejectedValue(new Error('Notification failed'));

        const mockSetFailed = jest.spyOn(core, 'setFailed');

        await run();

        expect(mockSetFailed).toHaveBeenCalledWith(expect.any(String));
    });

    it('should allow raw payload', async () => {
        const params = {
            ...defaultParams,
            raw: JSON.stringify({ custom: 'payload' })
        };
        core.getInput.mockImplementation((name) => params[name]);
        const mockNotify = jest.spyOn(MSTeams.prototype, 'notify').mockImplementation(jest.fn());

        await run();

        expect(mockNotify).toHaveBeenCalledWith('dummy_webhook', { custom: 'payload' });
    });
});

describe('run function with dry_run', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should skip notification when dry_run is true', async () => {
		core.getInput.mockImplementation((name) => {
			if (name === 'dry_run') return 'true';
			return '{}';
		});
		const mockNotify = jest.spyOn(MSTeams.prototype, 'notify').mockImplementation(jest.fn());
		const mockInfo = jest.spyOn(core, 'info');

		await run();

		expect(mockNotify).not.toHaveBeenCalled();
		expect(mockInfo).toHaveBeenCalledWith('Dry run - skipping notification send. Done.');
	});

	it('should send notification when dry_run is false', async () => {
		core.getInput.mockImplementation((name) => {
			if (name === 'dry_run') return 'false';
			if (name === 'webhook_url') return 'dummy_webhook';
			return '{}';
		});
		const mockNotify = jest.spyOn(MSTeams.prototype, 'notify').mockImplementation(jest.fn());
		const mockInfo = jest.spyOn(core, 'info');

		await run();

		expect(mockNotify).toHaveBeenCalled();
		expect(mockInfo).toHaveBeenCalledWith('Sent message to Microsoft Teams');
	});
});

describe('run function when webhook_url is missing', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should throw an error when webhook_url is missing', async () => {
		core.getInput.mockImplementation((name) => {
			if (name === 'webhook_url') return '';
			return '{}';
		});
		const mockSetFailed = jest.spyOn(core, 'setFailed');

		await run();

		expect(mockSetFailed).toHaveBeenCalledWith(
			'[Error] Missing Microsoft Teams Incoming Webhooks URL.\n' +
			'Please configure "MSTEAMS_WEBHOOK" as environment variable or\n' +
			'specify the key called "webhook_url" in "with" section.'
		);
	});
});


describe('run function with various combinations of job, steps, and needs', () => {
    beforeEach(() => {
        jest.restoreAllMocks();
    });

    const defaults = {
        description: 'default',
        job: { id: 'job_1', status: 'success' },
        steps: [{ name: 'step_1', status: 'success', outputs: { key1: 'value1' } }],
        needs: { need_1: { result: 'success', outputs: { key2: 'value2' } } },
        msteams_emails: ''
    }

    // Define combinations of job, steps, and needs with outputs
    const combinations = [
        {
            // all parameters are present
            ...defaults
        },
        {
            // job is missing
            ...defaults,
            job: null
        },
        {
            // steps are empty
            ...defaults,
            steps: []
        },
        {
            // needs are empty
            ...defaults,
            needs: {}
        },
        {
            // all parameters are empty
            description: null,
            job: null,
            steps: [],
            needs: {},
        },
        {
            // steps contain a mix of statuses with outputs
            ...defaults,
            steps: [
                { name: 'step_1', status: 'success', outputs: { key1: 'value1' } },
                { name: 'step_2', status: 'failure', outputs: { key3: 'value3' } }
            ]
        },
        {
            // needs contains unsupported status
            ...defaults,
            needs: {
                need_1: { result: 'custom', outputs: { key2: 'value2' } },
                need_2: { result: 'failure', outputs: { key4: 'value4' } },
                // check truncation of needs ids
                abcdefghijklmnoprstuwxyz: { result: 'success', outputs: { key5: 'value5' } },
            },
            msteams_emails: 'foo1@bar.com, foo2@bar.com'
        }
    ];

    const testPayloadRemainsUnchanged = async ({ description, job, steps, needs, msteams_emails }) => {
        const mockNotify = jest.spyOn(MSTeams.prototype, 'notify').mockImplementation(jest.fn());
    
        // Mock inputs based on the combination
        core.getInput.mockImplementation((name) => {
            switch (name) {
                case 'job': return job ? JSON.stringify(job) : '{}';
                case 'steps': return steps ? JSON.stringify(steps) : '[]';
                case 'needs': return needs ? JSON.stringify(needs) : '{}';
                case 'webhook_url': return 'dummy_webhook';
                case 'title': return 'Test Title';
                case 'msteams_emails': return msteams_emails;
                default: return '';
            }
        });
    
        await run();
    
        // Verify MSTeams.notify is called with the correct payload
        expect(mockNotify).toHaveBeenCalledWith('dummy_webhook', expect.any(Object));
    
        // Use a snapshot to ensure the payload remains consistent
        expect(mockNotify.mock.calls[0][1]).toMatchSnapshot('payload');
    };

    it.each(combinations)('should call MSTeams.notify with correct payload when %j', testPayloadRemainsUnchanged);

    it('should handle missing inputs', async () => {
        core.getInput.mockImplementation((name) => undefined);
        
        const mockSetFailed = jest.spyOn(core, 'setFailed');

        await run();

        expect(mockSetFailed).toHaveBeenCalledWith(expect.any(String));
    });
});
