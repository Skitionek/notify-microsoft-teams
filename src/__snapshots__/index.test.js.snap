// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run function with various combinations of job, steps, and needs should call MSTeams.notify with correct payload when {"description":"default","job":{"id":"job_1","status":"success"},"steps":[],"needs":{"need_1":{"result":"success","outputs":{"key2":"value2"}}},"msteams_emails":""}: payload 1`] = `
{
  "attachments": [
    {
      "content": {
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "body": [
          {
            "size": "Medium",
            "style": "heading",
            "text": "Test Title",
            "type": "TextBlock",
            "weight": "Bolder",
            "wrap": true,
          },
          {
            "size": "Medium",
            "text": "[undefined](html_url)",
            "type": "TextBlock",
            "weight": "lighter",
          },
          {
            "facts": [
              {
                "title": "✓ need_1",
                "value": "Success!",
              },
            ],
            "type": "FactSet",
          },
          {
            "columns": [
              {
                "items": [
                  {
                    "altText": "Result",
                    "size": "small",
                    "style": "person",
                    "type": "Image",
                    "url": "https://raw.githubusercontent.com/Skitionek/notify-microsoft-teams/master/icons/success.png",
                  },
                ],
                "type": "Column",
                "width": "auto",
              },
              {
                "items": [
                  {
                    "text": "Success!",
                    "type": "TextBlock",
                    "weight": "bolder",
                  },
                  {
                    "text": "timestamp",
                    "type": "TextBlock",
                  },
                ],
                "type": "Column",
                "width": "stretch",
              },
            ],
            "type": "ColumnSet",
          },
          {
            "actions": [
              {
                "title": "Repository",
                "type": "Action.OpenUrl",
                "url": "html_url",
              },
              {
                "title": "Compare",
                "type": "Action.OpenUrl",
                "url": "compare_url",
              },
            ],
            "type": "ActionSet",
          },
        ],
        "msteams": {
          "entities": [
            {},
          ],
        },
        "type": "AdaptiveCard",
        "version": "1.4",
      },
      "contentType": "application/vnd.microsoft.card.adaptive",
    },
  ],
  "type": "message",
}
`;

exports[`run function with various combinations of job, steps, and needs should call MSTeams.notify with correct payload when {"description":"default","job":{"id":"job_1","status":"success"},"steps":[{"name":"step_1","status":"success","outputs":{"key1":"value1"}},{"name":"step_2","status":"failure","outputs":{"key3":"value3"}}],"needs":{"need_1":{"result":"success","outputs":{"key2":"value2"}}},"msteams_emails":""}: payload 1`] = `
{
  "attachments": [
    {
      "content": {
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "body": [
          {
            "size": "Medium",
            "style": "heading",
            "text": "Test Title",
            "type": "TextBlock",
            "weight": "Bolder",
            "wrap": true,
          },
          {
            "size": "Medium",
            "text": "[undefined](html_url)",
            "type": "TextBlock",
            "weight": "lighter",
          },
          {
            "facts": [
              {
                "title": "? 0",
                "value": "No job context has been provided",
              },
              {
                "title": "? 1",
                "value": "No job context has been provided",
              },
            ],
            "type": "FactSet",
          },
          {
            "facts": [
              {
                "title": "✓ need_1",
                "value": "Success!",
              },
            ],
            "type": "FactSet",
          },
          {
            "columns": [
              {
                "items": [
                  {
                    "altText": "Result",
                    "size": "small",
                    "style": "person",
                    "type": "Image",
                    "url": "https://raw.githubusercontent.com/Skitionek/notify-microsoft-teams/master/icons/success.png",
                  },
                ],
                "type": "Column",
                "width": "auto",
              },
              {
                "items": [
                  {
                    "text": "Success!",
                    "type": "TextBlock",
                    "weight": "bolder",
                  },
                  {
                    "text": "timestamp",
                    "type": "TextBlock",
                  },
                ],
                "type": "Column",
                "width": "stretch",
              },
            ],
            "type": "ColumnSet",
          },
          {
            "actions": [
              {
                "title": "Repository",
                "type": "Action.OpenUrl",
                "url": "html_url",
              },
              {
                "title": "Compare",
                "type": "Action.OpenUrl",
                "url": "compare_url",
              },
            ],
            "type": "ActionSet",
          },
        ],
        "msteams": {
          "entities": [
            {},
          ],
        },
        "type": "AdaptiveCard",
        "version": "1.4",
      },
      "contentType": "application/vnd.microsoft.card.adaptive",
    },
  ],
  "type": "message",
}
`;

exports[`run function with various combinations of job, steps, and needs should call MSTeams.notify with correct payload when {"description":"default","job":{"id":"job_1","status":"success"},"steps":[{"name":"step_1","status":"success","outputs":{"key1":"value1"}}],"needs":{"need_1":{"result":"custom","outputs":{"key2":"value2"}},"need_2":{"result":"failure","outputs":{"key4":"value4"}},"abcdefghijklmnoprstuwxyz":{"result":"success","outputs":{"key5":"value5"}}},"msteams_emails":"foo1@bar.com, foo2@bar.com"}: payload 1`] = `
{
  "attachments": [
    {
      "content": {
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "body": [
          {
            "size": "Medium",
            "style": "heading",
            "text": "Test Title",
            "type": "TextBlock",
            "weight": "Bolder",
            "wrap": true,
          },
          {
            "size": "Medium",
            "text": "[undefined](html_url)",
            "type": "TextBlock",
            "weight": "lighter",
          },
          {
            "facts": [
              {
                "title": "? 0",
                "value": "No job context has been provided",
              },
            ],
            "type": "FactSet",
          },
          {
            "facts": [
              {
                "title": "? need_1",
                "value": "No job context has been provided",
              },
              {
                "title": "✗ need_2",
                "value": "Failure",
              },
              {
                "title": "✓ abcdefghijkl...",
                "value": "Success!",
              },
            ],
            "type": "FactSet",
          },
          {
            "columns": [
              {
                "items": [
                  {
                    "altText": "Result",
                    "size": "small",
                    "style": "person",
                    "type": "Image",
                    "url": "https://raw.githubusercontent.com/Skitionek/notify-microsoft-teams/master/icons/success.png",
                  },
                ],
                "type": "Column",
                "width": "auto",
              },
              {
                "items": [
                  {
                    "text": "Success!",
                    "type": "TextBlock",
                    "weight": "bolder",
                  },
                  {
                    "text": "timestamp",
                    "type": "TextBlock",
                  },
                ],
                "type": "Column",
                "width": "stretch",
              },
            ],
            "type": "ColumnSet",
          },
          {
            "actions": [
              {
                "title": "Repository",
                "type": "Action.OpenUrl",
                "url": "html_url",
              },
              {
                "title": "Compare",
                "type": "Action.OpenUrl",
                "url": "compare_url",
              },
            ],
            "type": "ActionSet",
          },
          {
            "text": "<at>foo1@bar.com</at> <at>foo2@bar.com</at>",
            "type": "TextBlock",
            "wrap": true,
          },
        ],
        "msteams": {
          "entities": [
            {
              "mentioned": {
                "id": "foo1@bar.com",
                "name": "foo1@bar.com",
              },
              "text": "<at>foo1@bar.com</at>",
              "type": "mention",
            },
            {
              "mentioned": {
                "id": "foo2@bar.com",
                "name": "foo2@bar.com",
              },
              "text": "<at>foo2@bar.com</at>",
              "type": "mention",
            },
          ],
        },
        "type": "AdaptiveCard",
        "version": "1.4",
      },
      "contentType": "application/vnd.microsoft.card.adaptive",
    },
  ],
  "type": "message",
}
`;

exports[`run function with various combinations of job, steps, and needs should call MSTeams.notify with correct payload when {"description":"default","job":{"id":"job_1","status":"success"},"steps":[{"name":"step_1","status":"success","outputs":{"key1":"value1"}}],"needs":{"need_1":{"result":"success","outputs":{"key2":"value2"}}},"msteams_emails":""}: payload 1`] = `
{
  "attachments": [
    {
      "content": {
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "body": [
          {
            "size": "Medium",
            "style": "heading",
            "text": "Test Title",
            "type": "TextBlock",
            "weight": "Bolder",
            "wrap": true,
          },
          {
            "size": "Medium",
            "text": "[undefined](html_url)",
            "type": "TextBlock",
            "weight": "lighter",
          },
          {
            "facts": [
              {
                "title": "? 0",
                "value": "No job context has been provided",
              },
            ],
            "type": "FactSet",
          },
          {
            "facts": [
              {
                "title": "✓ need_1",
                "value": "Success!",
              },
            ],
            "type": "FactSet",
          },
          {
            "columns": [
              {
                "items": [
                  {
                    "altText": "Result",
                    "size": "small",
                    "style": "person",
                    "type": "Image",
                    "url": "https://raw.githubusercontent.com/Skitionek/notify-microsoft-teams/master/icons/success.png",
                  },
                ],
                "type": "Column",
                "width": "auto",
              },
              {
                "items": [
                  {
                    "text": "Success!",
                    "type": "TextBlock",
                    "weight": "bolder",
                  },
                  {
                    "text": "timestamp",
                    "type": "TextBlock",
                  },
                ],
                "type": "Column",
                "width": "stretch",
              },
            ],
            "type": "ColumnSet",
          },
          {
            "actions": [
              {
                "title": "Repository",
                "type": "Action.OpenUrl",
                "url": "html_url",
              },
              {
                "title": "Compare",
                "type": "Action.OpenUrl",
                "url": "compare_url",
              },
            ],
            "type": "ActionSet",
          },
        ],
        "msteams": {
          "entities": [
            {},
          ],
        },
        "type": "AdaptiveCard",
        "version": "1.4",
      },
      "contentType": "application/vnd.microsoft.card.adaptive",
    },
  ],
  "type": "message",
}
`;

exports[`run function with various combinations of job, steps, and needs should call MSTeams.notify with correct payload when {"description":"default","job":{"id":"job_1","status":"success"},"steps":[{"name":"step_1","status":"success","outputs":{"key1":"value1"}}],"needs":{},"msteams_emails":""}: payload 1`] = `
{
  "attachments": [
    {
      "content": {
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "body": [
          {
            "size": "Medium",
            "style": "heading",
            "text": "Test Title",
            "type": "TextBlock",
            "weight": "Bolder",
            "wrap": true,
          },
          {
            "size": "Medium",
            "text": "[undefined](html_url)",
            "type": "TextBlock",
            "weight": "lighter",
          },
          {
            "facts": [
              {
                "title": "? 0",
                "value": "No job context has been provided",
              },
            ],
            "type": "FactSet",
          },
          {
            "columns": [
              {
                "items": [
                  {
                    "altText": "Result",
                    "size": "small",
                    "style": "person",
                    "type": "Image",
                    "url": "https://raw.githubusercontent.com/Skitionek/notify-microsoft-teams/master/icons/success.png",
                  },
                ],
                "type": "Column",
                "width": "auto",
              },
              {
                "items": [
                  {
                    "text": "Success!",
                    "type": "TextBlock",
                    "weight": "bolder",
                  },
                  {
                    "text": "timestamp",
                    "type": "TextBlock",
                  },
                ],
                "type": "Column",
                "width": "stretch",
              },
            ],
            "type": "ColumnSet",
          },
          {
            "actions": [
              {
                "title": "Repository",
                "type": "Action.OpenUrl",
                "url": "html_url",
              },
              {
                "title": "Compare",
                "type": "Action.OpenUrl",
                "url": "compare_url",
              },
            ],
            "type": "ActionSet",
          },
        ],
        "msteams": {
          "entities": [
            {},
          ],
        },
        "type": "AdaptiveCard",
        "version": "1.4",
      },
      "contentType": "application/vnd.microsoft.card.adaptive",
    },
  ],
  "type": "message",
}
`;

exports[`run function with various combinations of job, steps, and needs should call MSTeams.notify with correct payload when {"description":"default","job":null,"steps":[{"name":"step_1","status":"success","outputs":{"key1":"value1"}}],"needs":{"need_1":{"result":"success","outputs":{"key2":"value2"}}},"msteams_emails":""}: payload 1`] = `
{
  "attachments": [
    {
      "content": {
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "body": [
          {
            "size": "Medium",
            "style": "heading",
            "text": "Test Title",
            "type": "TextBlock",
            "weight": "Bolder",
            "wrap": true,
          },
          {
            "size": "Medium",
            "text": "[undefined](html_url)",
            "type": "TextBlock",
            "weight": "lighter",
          },
          {
            "facts": [
              {
                "title": "? 0",
                "value": "No job context has been provided",
              },
            ],
            "type": "FactSet",
          },
          {
            "facts": [
              {
                "title": "✓ need_1",
                "value": "Success!",
              },
            ],
            "type": "FactSet",
          },
          {
            "columns": [
              {
                "items": [
                  {
                    "altText": "Result",
                    "size": "small",
                    "style": "person",
                    "type": "Image",
                    "url": "https://raw.githubusercontent.com/Skitionek/notify-microsoft-teams/master/icons/unknown.png",
                  },
                ],
                "type": "Column",
                "width": "auto",
              },
              {
                "items": [
                  {
                    "text": "No job context has been provided",
                    "type": "TextBlock",
                    "weight": "bolder",
                  },
                  {
                    "text": "timestamp",
                    "type": "TextBlock",
                  },
                ],
                "type": "Column",
                "width": "stretch",
              },
            ],
            "type": "ColumnSet",
          },
          {
            "actions": [
              {
                "title": "Repository",
                "type": "Action.OpenUrl",
                "url": "html_url",
              },
              {
                "title": "Compare",
                "type": "Action.OpenUrl",
                "url": "compare_url",
              },
            ],
            "type": "ActionSet",
          },
        ],
        "msteams": {
          "entities": [
            {},
          ],
        },
        "type": "AdaptiveCard",
        "version": "1.4",
      },
      "contentType": "application/vnd.microsoft.card.adaptive",
    },
  ],
  "type": "message",
}
`;

exports[`run function with various combinations of job, steps, and needs should call MSTeams.notify with correct payload when {"description":null,"job":null,"steps":[],"needs":{}}: payload 1`] = `
{
  "attachments": [
    {
      "content": {
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "body": [
          {
            "size": "Medium",
            "style": "heading",
            "text": "Test Title",
            "type": "TextBlock",
            "weight": "Bolder",
            "wrap": true,
          },
          {
            "size": "Medium",
            "text": "[undefined](html_url)",
            "type": "TextBlock",
            "weight": "lighter",
          },
          {
            "columns": [
              {
                "items": [
                  {
                    "altText": "Result",
                    "size": "small",
                    "style": "person",
                    "type": "Image",
                    "url": "https://raw.githubusercontent.com/Skitionek/notify-microsoft-teams/master/icons/unknown.png",
                  },
                ],
                "type": "Column",
                "width": "auto",
              },
              {
                "items": [
                  {
                    "text": "No job context has been provided",
                    "type": "TextBlock",
                    "weight": "bolder",
                  },
                  {
                    "text": "timestamp",
                    "type": "TextBlock",
                  },
                ],
                "type": "Column",
                "width": "stretch",
              },
            ],
            "type": "ColumnSet",
          },
          {
            "actions": [
              {
                "title": "Repository",
                "type": "Action.OpenUrl",
                "url": "html_url",
              },
              {
                "title": "Compare",
                "type": "Action.OpenUrl",
                "url": "compare_url",
              },
            ],
            "type": "ActionSet",
          },
        ],
        "msteams": {
          "entities": [
            {},
          ],
        },
        "type": "AdaptiveCard",
        "version": "1.4",
      },
      "contentType": "application/vnd.microsoft.card.adaptive",
    },
  ],
  "type": "message",
}
`;
